File: sdk/go/tipsyabconfig/exposure.go
package tipsyabconfig

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	abtestv1 "github.com/Lightspeed-Intelligence/tipsy-ab-config-sdk/api/gen/go/tipsy/abtest/v1"
)

// ExposureEvent is the SDK-emitted exposure record. One event corresponds to
// one (uid, ns, key, version) hit through an abtest path. The shape is a
// SDK-stable struct so production callers can route to Kafka / HTTP / log
// without depending on the proto types.
type ExposureEvent struct {
	UserID           string
	Namespace        string
	Key              string
	Version          int64
	Source           string // "experiment_group" | "whitelist"
	ExperimentID     string // "" when source = whitelist (v2: TEXT ids)
	GroupID          string // "" when source = whitelist
	ExperimentStatus string // for source = experiment_group
	ReleaseID        int64  // 0 when source = experiment_group
	// TraceID is the per-request trace id of the AbtestContext that produced
	// this exposure (sdk-trace-id §4). It is the reporting hook for the
	// upcoming "experiment-result data report" channel: same id as the
	// server-side compute log so log ↔ report ↔ upstream caller all join on
	// one column.
	TraceID   string
	EmittedAt time.Time
}

// ExposureSink receives exposure events. Implementations should be
// non-blocking; the SDK calls Sink on a dedicated background goroutine.
type ExposureSink interface {
	Sink(ev ExposureEvent)
}

// ExposureSinkFunc adapts a plain function into an ExposureSink.
type ExposureSinkFunc func(ExposureEvent)

// Sink implements ExposureSink.
func (f ExposureSinkFunc) Sink(ev ExposureEvent) { f(ev) }

// logSink is the default sink — emits one INFO log per exposure.
type logSink struct{ logger *slog.Logger }

// Sink implements ExposureSink.
func (s *logSink) Sink(ev ExposureEvent) {
	s.logger.Info("tipsyabconfig: exposure",
		"uid", ev.UserID,
		"ns", ev.Namespace,
		"key", ev.Key,
		"version", ev.Version,
		"source", ev.Source,
		"experiment_id", ev.ExperimentID,
		"group_id", ev.GroupID,
		"release_id", ev.ReleaseID,
		"trace_id", ev.TraceID,
	)
}

// exposureEmitter is the SDK's async exposure pipeline. emit() is
// non-blocking and offers a per-process LRU-style dedup window keyed on
// (uid, key, version) per design §9.2.
//
// Implementation: an unbuffered map[string]time.Time guarded by RWMutex
// holds last-seen timestamps. On emit we drop the event if a recent entry
// exists; otherwise we push onto a buffered channel and the background
// goroutine drains to the sink. We GC the dedup map opportunistically when
// it grows past dedupMaxEntries.
type exposureEmitter struct {
	sink ExposureSink
	ttl  time.Duration

	logger *slog.Logger

	queue chan ExposureEvent

	dmu             sync.Mutex
	dedup           map[string]time.Time
	dedupMaxEntries int
	now             func() time.Time
}

const defaultExposureQueueSize = 4096

func newExposureEmitter(sink ExposureSink, ttl time.Duration, logger *slog.Logger) *exposureEmitter {
	if logger == nil {
		logger = slog.Default()
	}
	if sink == nil {
		sink = &logSink{logger: logger}
	}
	return &exposureEmitter{
		sink:            sink,
		ttl:             ttl,
		logger:          logger,
		queue:           make(chan ExposureEvent, defaultExposureQueueSize),
		dedup:           make(map[string]time.Time, 1024),
		dedupMaxEntries: 16384,
		now:             time.Now,
	}
}

// emit is the hot-path entry called from GetConfig. It builds zero-or-more
// ExposureEvent rows (one per hit-eligible exposure on the matching key),
// deduplicates each, and posts to the background queue.
//
// abExposures is the Exposure list returned by AbtestService.Compute. We
// only emit a row when the exposure's (key, version) match the resolved
// (key, version) — the response can carry hits for many keys but the SDK
// only attributes the one we actually returned.
//
// traceID is the AbtestContext's per-request trace id (sdk-trace-id §4); it
// is stamped onto every emitted event so the log sink and the upcoming
// upstream "experiment-result data report" channel can join on the same id
// as the server-side compute log.
func (e *exposureEmitter) emit(uid, ns, key string, version int64, abExposures []*abtestv1.Exposure, traceID string) {
	if e == nil {
		return
	}
	matched := pickExposureForKey(key, version, abExposures)
	if matched == nil {
		// No source attribution in the Compute response; per design we
		// still record the hit at "experiment_group" with zero ids so
		// downstream dashboards know an SDK-side resolve happened.
		matched = &abtestv1.Exposure{Key: key, Version: version, Source: "experiment_group"}
	}
	ev := ExposureEvent{
		UserID:    uid,
		Namespace: ns,
		Key:       key,
		Version:   version,
		Source:    matched.GetSource(),
		TraceID:   traceID,
		EmittedAt: e.now(),
	}
	if matched.ExperimentId != nil {
		ev.ExperimentID = *matched.ExperimentId
	}
	if matched.GroupId != nil {
		ev.GroupID = *matched.GroupId
	}
	if matched.ExperimentStatus != nil {
		ev.ExperimentStatus = *matched.ExperimentStatus
	}
	if matched.ReleaseId != nil {
		ev.ReleaseID = *matched.ReleaseId
	}
	if !e.shouldEmit(uid, key, version) {
		return
	}
	select {
	case e.queue <- ev:
	default:
		// Queue full — drop with a warn. We do not block GetConfig.
		e.logger.Warn("tipsyabconfig: exposure queue full; dropping event",
			"ns", ns, "key", key, "version", version)
	}
}

func pickExposureForKey(key string, version int64, exposures []*abtestv1.Exposure) *abtestv1.Exposure {
	for _, ex := range exposures {
		if ex == nil {
			continue
		}
		if ex.Key == key && ex.Version == version {
			return ex
		}
	}
	return nil
}

// shouldEmit returns true if this (uid, key, version) tuple has not been
// seen within the ttl window. Returning false means we drop the event as a
// duplicate. We always update the last-seen timestamp when returning true.
func (e *exposureEmitter) shouldEmit(uid, key string, version int64) bool {
	k := dedupKey(uid, key, version)
	now := e.now()
	e.dmu.Lock()
	defer e.dmu.Unlock()
	if last, ok := e.dedup[k]; ok && now.Sub(last) < e.ttl {
		return false
	}
	if len(e.dedup) >= e.dedupMaxEntries {
		// Opportunistic GC: drop entries past their TTL.
		for ek, ev := range e.dedup {
			if now.Sub(ev) >= e.ttl {
				delete(e.dedup, ek)
			}
		}
	}
	e.dedup[k] = now
	return true
}

func dedupKey(uid, key string, version int64) string {
	return fmt.Sprintf("%s\x1f%s\x1f%d", uid, key, version)
}

// run drains the exposure queue. Exits when ctx is cancelled and the queue
// is empty.
func (e *exposureEmitter) run(ctx context.Context, wg *sync.WaitGroup) {
	defer wg.Done()
	for {
		select {
		case <-ctx.Done():
			// Drain any pending events synchronously, then exit.
			for {
				select {
				case ev := <-e.queue:
					e.sink.Sink(ev)
				default:
					return
				}
			}
		case ev := <-e.queue:
			func() {
				defer func() {
					if r := recover(); r != nil {
						e.logger.Error("tipsyabconfig: exposure sink panicked", "panic", r)
					}
				}()
				e.sink.Sink(ev)
			}()
		}
	}
}

File: sdk/python/tipsy_ab_config/exposure.py
"""Asynchronous exposure emitter with 5-minute per-process dedup.

Mirrors ``sdk/go/tipsyabconfig/exposure.go``. The Python implementation is
asyncio-native: one background task drains a bounded ``asyncio.Queue`` and
hands events to the registered :class:`ExposureSink`. The dedup map is
``dict[str, float]`` guarded by the queue's coroutine loop (no thread lock —
exposures only enter from coroutine code paths reachable from
``get_config``).
"""

from __future__ import annotations

import asyncio
import logging
import time
from dataclasses import dataclass, field
from typing import Awaitable, Callable, Iterable, List, Optional, Protocol

logger = logging.getLogger("tipsy_ab_config.exposure")


@dataclass
class ExposureEvent:
    """One (uid, ns, key, version) hit attributed to an abtest path."""

    user_id: str
    namespace: str
    key: str
    version: int
    source: str = ""  # "experiment_group" | "whitelist"
    experiment_id: str = ""  # v2: TEXT ids; "" when source = whitelist
    group_id: str = ""  # v2: TEXT ids; "" when source = whitelist
    experiment_status: str = ""
    release_id: int = 0  # gray release_whitelist.id stays int64
    emitted_at: float = field(default_factory=time.time)


class ExposureSink(Protocol):
    """Sink receives one event per exposure.

    Implementations may be async (``await sink(ev)`` is awaited) — the
    emitter inspects the return type and awaits when needed.
    """

    def __call__(self, ev: ExposureEvent) -> Optional[Awaitable[None]]:
        ...


def _default_sink(ev: ExposureEvent) -> None:
    logger.info(
        "exposure",
        extra={
            "uid": ev.user_id,
            "ns": ev.namespace,
            "key": ev.key,
            "version": ev.version,
            "source": ev.source,
            "experiment_id": ev.experiment_id,
            "group_id": ev.group_id,
            "release_id": ev.release_id,
        },
    )


DEFAULT_QUEUE_SIZE = 4096
DEFAULT_DEDUP_MAX_ENTRIES = 16384


class ExposureEmitter:
    """Async fire-and-forget exposure pipeline.

    Use:

        emitter = ExposureEmitter(sink, ttl_seconds=300)
        await emitter.start()              # spawns background task
        emitter.emit(uid, ns, key, ver, ab_exposures)
        ...
        await emitter.aclose()             # drains queue, exits task
    """

    def __init__(
        self,
        sink: Optional[ExposureSink] = None,
        ttl_seconds: float = 300.0,
        queue_size: int = DEFAULT_QUEUE_SIZE,
        dedup_max_entries: int = DEFAULT_DEDUP_MAX_ENTRIES,
        loop: Optional[asyncio.AbstractEventLoop] = None,
        clock: Callable[[], float] = time.time,
    ) -> None:
        self._sink: ExposureSink = sink or _default_sink
        self._ttl = ttl_seconds
        self._queue: asyncio.Queue[Optional[ExposureEvent]] = asyncio.Queue(
            maxsize=queue_size
        )
        self._dedup: dict[str, float] = {}
        self._dedup_max = dedup_max_entries
        self._loop = loop
        self._clock = clock
        self._task: Optional[asyncio.Task[None]] = None
        self._closed = False

    async def start(self) -> None:
        if self._task is not None:
            return
        self._loop = self._loop or asyncio.get_running_loop()
        self._task = self._loop.create_task(self._run())

    def emit(
        self,
        user_id: str,
        namespace: str,
        key: str,
        version: int,
        ab_exposures: Iterable,
    ) -> None:
        """Non-blocking emit. Drops the event on a full queue."""
        if self._closed:
            return
        matched = _pick_exposure_for_key(key, version, ab_exposures)
        if matched is None:
            # No source attribution from Compute (shouldn't happen for an
            # abtest hit, but stay defensive). Default to experiment_group
            # with zero ids so downstream still records the resolve.
            ev = ExposureEvent(
                user_id=user_id,
                namespace=namespace,
                key=key,
                version=version,
                source="experiment_group",
                emitted_at=self._clock(),
            )
        else:
            ev = ExposureEvent(
                user_id=user_id,
                namespace=namespace,
                key=key,
                version=version,
                source=matched.source,
                experiment_id=matched.experiment_id if matched.HasField("experiment_id") else "",
                group_id=matched.group_id if matched.HasField("group_id") else "",
                experiment_status=matched.experiment_status if matched.HasField("experiment_status") else "",
                release_id=int(matched.release_id) if matched.HasField("release_id") else 0,
                emitted_at=self._clock(),
            )
        if not self._should_emit(user_id, key, version):
            return
        try:
            self._queue.put_nowait(ev)
        except asyncio.QueueFull:
            logger.warning(
                "tipsy_ab_config: exposure queue full; dropping event ns=%s key=%s version=%d",
                namespace,
                key,
                version,
            )

    def _should_emit(self, uid: str, key: str, version: int) -> bool:
        now = self._clock()
        k = f"{uid}\x1f{key}\x1f{version}"
        last = self._dedup.get(k)
        if last is not None and (now - last) < self._ttl:
            return False
        if len(self._dedup) >= self._dedup_max:
            # Opportunistic GC.
            stale = [ek for ek, t in self._dedup.items() if (now - t) >= self._ttl]
            for ek in stale:
                del self._dedup[ek]
        self._dedup[k] = now
        return True

    async def _run(self) -> None:
        while True:
            ev = await self._queue.get()
            if ev is None:
                return
            try:
                result = self._sink(ev)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:  # noqa: BLE001
                logger.exception("tipsy_ab_config: exposure sink raised")

    async def aclose(self) -> None:
        """Signal the worker to drain and exit, then await it."""
        if self._closed:
            return
        self._closed = True
        # Sentinel.
        try:
            self._queue.put_nowait(None)
        except asyncio.QueueFull:
            # Best-effort: if full, the worker is far behind. Cancel.
            if self._task is not None:
                self._task.cancel()
        if self._task is not None:
            try:
                await self._task
            except (asyncio.CancelledError, Exception):  # noqa: BLE001
                pass


def _pick_exposure_for_key(key: str, version: int, exposures: Iterable):
    for ex in exposures:
        if ex is None:
            continue
        if ex.key == key and int(ex.version) == int(version):
            return ex
    return None
